src/Psalm/Internal/Analyzer/Statements/Expression/SimpleTypeInferer.php

```diff
@@ -547,4 +413,200 @@ public static function infer(
 
         return null;
     }
+
+    /**
+     * @param   ?array<string, ClassConstantStorage> $existing_class_constants
+     */
+    private static function inferArrayType(
+        \Psalm\Codebase $codebase,
+        \Psalm\Internal\Provider\NodeDataProvider $nodes,
+        PhpParser\Node\Expr\Array_ $stmt,
+        \Psalm\Aliases $aliases,
+        \Psalm\FileSource $file_source = null,
+        ?array $existing_class_constants = null,
+        ?string $fq_classlike_name = null
+    ): ?Type\Union {
+        if (count($stmt->items) === 0) {
+            return Type::getEmptyArray();
+        }
+
+        $array_creation_info = new ArrayCreationInfo();
+
+        foreach ($stmt->items as $int_offset => $item) {
+            if ($item === null) {
+                continue;
+            }
+
+            if (!self::handleArrayItem(
+                $codebase,
+                $nodes,
+                $array_creation_info,
+                $int_offset,
+                $item,
+                $aliases,
+                $file_source,
+                $existing_class_constants,
+                $fq_classlike_name
+            )) {
+                return null;
+            }
+        }
+
+        $item_key_type = null;
+        if ($array_creation_info->item_key_atomic_types) {
+            $item_key_type = TypeCombiner::combine(
+                $array_creation_info->item_key_atomic_types,
+                null,
+                false,
+                true,
+                30
+            );
+        }
+
+        $item_value_type = null;
+        if ($array_creation_info->item_value_atomic_types) {
+            $item_value_type = TypeCombiner::combine(
+                $array_creation_info->item_value_atomic_types,
+                null,
+                false,
+                true,
+                30
+            );
+        }
+
+        // if this array looks like an object-like array, let's return that instead
```

Comment:
Maybe a similar check could be done for inferring TList?

Reply:
You mean, for cases where it crossed the threshold for `can_create_objectlike`?

Reply:
I saw multiple cases where can_create_objectlike can be false, so any of those I guess